### PR TITLE
[GLUTEN-5630][VL] Decrease peak memory by taking freeBytes into account 

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

After this PR, the peak memory usage decreased from 1.1G to 816M, and add ListenableMemoryAllocator into memory usage dump.

```
24/05/07 16:56:14 WARN [Executor task launch worker for task 1.0 in stage 7.0 (TID 1025)] NativeMemoryManager: About to release memory manager, usage dump:
	ShuffleWriter:                    Current used bytes: 1024.0 KiB, peak bytes:  212.3 MiB
	+- ShuffleWriter_root:            Current used bytes:      0.0 B, peak bytes: 1024.0 KiB
	|  \- ShuffleWriter_default_leaf: Current used bytes:      0.0 B, peak bytes:   1408.0 B
	\- gluten_allocator:              Current used bytes:      0.0 B, peak bytes:  211.3 MiB
```

(Fixes: \#5630)

### CODE REVIEW

File: cpp/core/jni/JniCommon.h [hunk 1]
```diff
@@ -368,10 +374,10 @@ class SparkAllocationListener final : public gluten::AllocationListener {
     if (granted < 0) {
       env->CallLongMethod(jListenerGlobalRef_, jUnreserveMethod_, -granted);
       checkException(env);
-      return;
+    } else {
+      env->CallLongMethod(jListenerGlobalRef_, jReserveMethod_, granted);
+      checkException(env);
     }
```
Comment (zhztheplayer) [inline cpp/core/jni/JniCommon.h:380]: Is this change all about code style?

**Comment by Yohahaha** [comment]
before
```
24/05/09 10:01:50 DEBUG [Executor task launch worker for task 11.0 in stage 7.0 (TID 1035)] NativeMemoryManager: About to release memory manager, usage dump:
	ShuffleWriter:                    Current used bytes: 1024.0 KiB, peak bytes:  201.6 MiB
	+- ShuffleWriter_root:            Current used bytes:      0.0 B, peak bytes: 1024.0 KiB
	|  \- ShuffleWriter_default_leaf: Current used bytes:      0.0 B, peak bytes:   1408.0 B
	\- gluten_allocator:              Current used bytes:      0.0 B, peak bytes:  200.6 MiB
24/05/09 10:01:50 DEBUG [Executor task launch worker for task 11.0 in stage 7.0 (TID 1035)] NativeMemoryManager: About to release memory manager, usage dump:
	WholeStageIterator:                      Current used bytes: 917.0 MiB, peak bytes:  917.0 MiB
	+- WholeStageIterator_root:              Current used bytes:     0.0 B, peak bytes:  577.0 MiB
	|  +- task.Gluten_Stage_7_TID_1035:      Current used bytes:     0.0 B, peak bytes:  577.0 MiB
	|  |  +- node.2:                         Current used bytes:     0.0 B, peak bytes:   32.0 MiB
	|  |  |  +- op.2.1.0.HashBuild:          Current used bytes:     0.0 B, peak bytes:    6.1 MiB
	|  |  |  \- op.2.0.0.HashProbe:          Current used bytes:     0.0 B, peak bytes:  675.0 KiB
	|  |  +- node.4:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  |  \- op.4.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  +- node.1:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  |  \- op.1.1.0.ValueStream:        Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  +- node.7:                         Current used bytes:     0.0 B, peak bytes:  552.0 MiB
	|  |  |  \- op.7.0.0.PartialAggregation: Current used bytes:     0.0 B, peak bytes:  493.5 MiB
	|  |  +- node.8:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  |  \- op.8.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  +- node.5:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	|  |  |  \- op.5.0.0.Expand:             Current used bytes:     0.0 B, peak bytes:   1024.0 B
	|  |  +- node.9:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	|  |  |  \- op.9.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:   24.0 KiB
	|  |  +- node.0:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  |  \- op.0.0.0.ValueStream:        Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |  +- node.6:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	|  |  |  \- op.6.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:   97.5 KiB
	|  |  \- node.3:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  |     \- op.3.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	|  \- WholeStageIterator_default_leaf:   Current used bytes:     0.0 B, peak bytes:      0.0 B
	\- gluten_allocator:                     Current used bytes:     0.0 B, peak bytes:      0.0 B
```
after
```
24/05/09 09:53:28 DEBUG [Executor task launch worker for task 11.0 in stage 7.0 (TID 1035)] NativeMemoryManager: About to release memory manager, usage dump:
	ShuffleWriter:                    Current used bytes: 1024.0 KiB, peak bytes:  202.5 MiB
	+- ShuffleWriter_root:            Current used bytes:      0.0 B, peak bytes: 1024.0 KiB
	|  \- ShuffleWriter_default_leaf: Current used bytes:      0.0 B, peak bytes:   1408.0 B
	\- gluten_allocator:              Current used bytes:      0.0 B, peak bytes:  201.5 MiB
24/05/09 09:53:28 DEBUG [Executor task launch worker for task 11.0 in stage 7.0 (TID 1035)] NativeMemoryManager: About to release memory manager, usage dump:
	WholeStageIterator:                      Current used bytes: 577.0 MiB, peak bytes:  577.0 MiB
	+- gluten_allocator:                     Current used bytes:     0.0 B, peak bytes:      0.0 B
	\- WholeStageIterator_root:              Current used bytes:     0.0 B, peak bytes:  577.0 MiB
	   +- WholeStageIterator_default_leaf:   Current used bytes:     0.0 B, peak bytes:      0.0 B
	   \- task.Gluten_Stage_7_TID_1035:      Current used bytes:     0.0 B, peak bytes:  577.0 MiB
	      +- node.3:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	      |  \- op.3.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	      +- node.0:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	      |  \- op.0.0.0.ValueStream:        Current used bytes:     0.0 B, peak bytes:      0.0 B
	      +- node.9:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	      |  \- op.9.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:   24.0 KiB
	      +- node.7:                         Current used bytes:     0.0 B, peak bytes:  552.0 MiB
	      |  \- op.7.0.0.PartialAggregation: Current used bytes:     0.0 B, peak bytes:  495.5 MiB
	      +- node.6:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	      |  \- op.6.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:   97.5 KiB
	      +- node.5:                         Current used bytes:     0.0 B, peak bytes: 1024.0 KiB
	      |  \- op.5.0.0.Expand:             Current used bytes:     0.0 B, peak bytes:   1024.0 B
	      +- node.2:                         Current used bytes:     0.0 B, peak bytes:   32.0 MiB
	      |  +- op.2.0.0.HashProbe:          Current used bytes:     0.0 B, peak bytes:  675.0 KiB
	      |  \- op.2.1.0.HashBuild:          Current used bytes:     0.0 B, peak bytes:    6.1 MiB
	      +- node.8:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	      |  \- op.8.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	      +- node.4:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	      |  \- op.4.0.0.FilterProject:      Current used bytes:     0.0 B, peak bytes:      0.0 B
	      \- node.1:                         Current used bytes:     0.0 B, peak bytes:      0.0 B
	         \- op.1.1.0.ValueStream:        Current used bytes:     0.0 B, peak bytes:      0.0 B
```

File: cpp/velox/memory/VeloxMemoryManager.cc [hunk 2]
```diff
@@ -208,7 +215,13 @@ int64_t shrinkVeloxMemoryPool(velox::memory::MemoryManager* mm, velox::memory::M
 } // namespace
 
 const MemoryUsageStats VeloxMemoryManager::collectMemoryUsageStats() const {
-  return collectMemoryUsageStatsInternal(veloxAggregatePool_.get());
+  MemoryUsageStats stats;
+  stats.set_current(listener_->currentBytes());
+  stats.set_peak(listener_->peakBytes());
+  stats.mutable_children()->emplace("gluten_allocator", collectGlutenAllocatorMemoryUsageStats(glutenAlloc_.get()));
```
Comment (zhztheplayer) [inline cpp/velox/memory/VeloxMemoryManager.cc:221]: Nit: Can it be better to use name `"gluten::MemoryAllocator"`? We have so many `allocator` definitions so it's better to use a very specific identifier to distinguish with others.

Comment (Yohahaha) [reply cpp/velox/memory/VeloxMemoryManager.cc]: ok.

**Review by zhztheplayer — APPROVED** [review]
Thanks for the fix!

**Pull request merged**